src/client.rs

```diff
@@ -106,54 +110,102 @@ impl Client {
+    ///
+    /// Returns Ok if the track was retreived and written successfully, an error otherwise.
+    ///
+    /// Note:
+    ///   futures::io::AsyncRead is not compatible with tokio::io::AsyncRead.
+    ///   and hence the need for a compatibility layer between tokio and futures
+    ///   which is in the tokio-util crate
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use soundcloud::Client;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///   let client = Client::new(env!("SOUNDCLOUD_CLIENT_ID"));
+    ///   let path = Path::new("hi.mp3");
+    ///   let track = client.tracks().id(263801976).get().await.unwrap();
+    ///   let mut outfile = tokio::fs::File::create(path).await.unwrap();
+    ///   let num_bytes = client.download(&track, &mut outfile).await.unwrap();
+    ///   assert!(num_bytes > 0);
+    /// }
+    /// ```
+    pub async fn download<W: AsyncWrite + Unpin>(&self, track: &Track, mut writer: W) -> Result<u64> {
+        if !track.downloadable || !track.download_url.is_some() {
+            return Err(Error::TrackNotDownloadable);
+        }
+        let url = self.parse_url(track.download_url.as_ref().unwrap());
+        let mut response = self.http_client.get(url).send().await?;
+        // Follow the redirect just this once.
+        if let Some(header) = response.headers().get(reqwest::header::LOCATION).cloned() {
+            let url = Url::parse(header.to_str()?).unwrap();
+            response = self.http_client.get(url).send().await?;
+        }
+        let download = response.bytes_stream();
+        // Convert the stream into an futures::io::AsyncRead
+        // but first we must convert the reqwest::Error into a futures::io::Error
+        let download = download.map_err(
+            |e| futures::io::Error::new(futures::io::ErrorKind::Other, e))
+            .into_async_read();
+        // Convert the futures::io::AsyncRead into a tokio::io::AsyncRead
+        let mut download = download.compat();
+        let bytes = tokio::io::copy(&mut download, &mut writer).await?;
+        Ok(bytes)
+    }
+
+    /// Starts streaming the track provided in the tracks `stream_url` to the `writer`
+    /// if the track is streamable via the API.
+    ///
+    /// Returns Ok if the track was retreived and written successfully, an error otherwise.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use soundcloud::Client;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///   let client = Client::new(env!("SOUNDCLOUD_CLIENT_ID"));
+    ///   let path = Path::new("hi.mp3");
+    ///   let track = client.tracks().id(263801976).get().await.unwrap();
+    ///   let mut outfile = tokio::fs::File::create(path).await.unwrap();
+    ///   let num_bytes = client.stream(&track, &mut outfile).await.unwrap();
+    ///   assert!(num_bytes > 0);
+    /// }
+    /// ```
+    pub async fn stream<W: AsyncWrite + Unpin>(&self, track: &Track, mut writer: W) -> Result<u64> {
```

Comment:
As the `download` and `stream` method are quite similar we should just have a helper like `read_url` which contains the actual http request and response streaming and just call it with `track.stream_url` or `track.download_url`.

Reply:
Agreed. This makes the caller of ```read_url``` responsible for checking ```track.streamable/track.downloadable```. Maybe wrap ```read_url``` with ```stream/download``` methods to perform this check?

Reply:
That was actually what I was suggesting. Provide read_url as an internal helper and keep the stream and download methods so they can do the required sanity checks.
Sorry if that wasn't clear.

Reply:
Done.